In [ ]:
import pandas as pd

In [363]:
df = pd.read_csv(r'D:\AOU-Graduation-Project\BioIntellect\AI\ECG\Data\ptb-xl-a-large-publicly-available-electrocardiography-dataset-1.0.3\ptbxl_database.csv', index_col='ecg_id')
df

,patient_id,age,sex,height,weight,nurse,site,device,recording_date,report,...,validated_by_human,baseline_drift,static_noise,burst_noise,electrodes_problems,extra_beats,pacemaker,strat_fold,filename_lr,filename_hr
ecg_id,,,,,,,,,,,,,,,,,,,,,
1,15709.0,56.0,1,NaN,63.0,2.0,0.0,CS-12 E,1984-11-09 09:17:34,sinusrhythmus periphere niederspannung,...,True,NaN,", I-V1,",NaN,NaN,NaN,NaN,3,records100/00000/00001_lr,records500/00000/00001_hr
2,13243.0,19.0,0,NaN,70.0,2.0,0.0,CS-12 E,1984-11-14 12:55:37,sinusbradykardie sonst normales ekg,...,True,NaN,NaN,NaN,NaN,NaN,NaN,2,records100/00000/00002_lr,records500/00000/00002_hr
3,20372.0,37.0,1,NaN,69.0,2.0,0.0,CS-12 E,1984-11-15 12:49:10,sinusrhythmus normales ekg,...,True,NaN,NaN,NaN,NaN,NaN,NaN,5,records100/00000/00003_lr,records500/00000/00003_hr
4,17014.0,24.0,0,NaN,82.0,2.0,0.0,CS-12 E,1984-11-15 13:44:57,sinusrhythmus normales ekg,...,True,", II,III,AVF",NaN,NaN,NaN,NaN,NaN,3,records100/00000/00004_lr,records500/00000/00004_hr
5,17448.0,19.0,1,NaN,70.0,2.0,0.0,CS-12 E,1984-11-17 10:43:15,sinusrhythmus normales ekg,...,True,", III,AVR,AVF",NaN,NaN,NaN,NaN,NaN,4,records100/00000/00005_lr,records500/00000/00005_hr
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
21833,17180.0,67.0,1,NaN,NaN,1.0,2.0,AT-60 3,2001-05-31 09:14:35,ventrikulÄre extrasystole(n) sinustachykardie ...,...,True,NaN,", alles,",NaN,NaN,1ES,NaN,7,records100/21000/21833_lr,records500/21000/21833_hr
21834,20703.0,300.0,0,NaN,NaN,1.0,2.0,AT-60 3,2001-06-05 11:33:39,sinusrhythmus lagetyp normal qrs(t) abnorm ...,...,True,NaN,NaN,NaN,NaN,NaN,NaN,4,records100/21000/21834_lr,records500/21000/21834_hr
21835,19311.0,59.0,1,NaN,NaN,1.0,2.0,AT-60 3,2001-06-08 10:30:27,sinusrhythmus lagetyp normal t abnorm in anter...,...,True,NaN,", I-AVR,",NaN,NaN,NaN,NaN,2,records100/21000/21835_lr,records500/21000/21835_hr


In [364]:
df['patient_id'].nunique()

18869

In [365]:
df.columns

Index(['patient_id', 'age', 'sex', 'height', 'weight', 'nurse', 'site',
       'device', 'recording_date', 'report', 'scp_codes', 'heart_axis',
       'infarction_stadium1', 'infarction_stadium2', 'validated_by',
       'second_opinion', 'initial_autogenerated_report', 'validated_by_human',
       'baseline_drift', 'static_noise', 'burst_noise', 'electrodes_problems',
       'extra_beats', 'pacemaker', 'strat_fold', 'filename_lr', 'filename_hr'],
      dtype='object')

In [366]:
df.describe()

,patient_id,age,sex,height,weight,nurse,site,validated_by,strat_fold
count,21799.000000,21799.000000,21799.000000,6974.000000,9421.000000,20326.000000,21782.000000,12421.000000,21799.000000
mean,11250.156521,62.769301,0.479150,166.702323,70.995223,2.291745,1.544945,0.746075,5.503005
std,6235.026404,32.308813,0.499577,10.867321,15.878803,3.254033,4.172883,1.178003,2.874948
min,302.000000,2.000000,0.000000,6.000000,5.000000,0.000000,0.000000,0.000000,1.000000
25%,5974.500000,50.000000,0.000000,160.000000,60.000000,0.000000,0.000000,0.000000,3.000000
50%,11419.000000,62.000000,0.000000,166.000000,70.000000,1.000000,1.000000,1.000000,6.000000
75%,16607.500000,72.000000,1.000000,174.000000,80.000000,3.000000,2.000000,1.000000,8.000000
max,21797.000000,300.000000,1.000000,209.000000,250.000000,11.000000,50.000000,11.000000,10.000000


In [367]:
df.isnull().sum()

patient_id                          0
age                                 0
sex                                 0
height                          14825
weight                          12378
nurse                            1473
site                               17
device                              0
recording_date                      0
report                              0
scp_codes                           0
heart_axis                       8468
infarction_stadium1             16187
infarction_stadium2             21696
validated_by                     9378
second_opinion                      0
initial_autogenerated_report        0
validated_by_human                  0
baseline_drift                  20201
static_noise                    18539
burst_noise                     21186
electrodes_problems             21769
extra_beats                     19850
pacemaker                       21508
strat_fold                          0
filename_lr                         0
filename_hr 

### Age preprocessing 

In [368]:
df["age"].value_counts(dropna=False)


age
67.0    558
68.0    553
65.0    551
66.0    537
54.0    530
       ... 
5.0       2
4.0       2
2.0       1
6.0       1
7.0       1
Name: count, Length: 89, dtype: int64

In [369]:
df["age"].max()

np.float64(300.0)

In [370]:
df[(df["age"] ==300) & (df["age"] >= 123)]

,patient_id,age,sex,height,weight,nurse,site,device,recording_date,report,...,validated_by_human,baseline_drift,static_noise,burst_noise,electrodes_problems,extra_beats,pacemaker,strat_fold,filename_lr,filename_hr
ecg_id,,,,,,,,,,,,,,,,,,,,,
108,11810.0,300.0,1,NaN,57.0,0.0,0.0,CS-12 E,1985-12-20 11:54:45,sinustachykardie sonst normales ekg,...,True,NaN,NaN,NaN,NaN,NaN,NaN,8,records100/00000/00108_lr,records500/00000/00108_hr
255,4867.0,300.0,1,163.0,58.0,10.0,1.0,AT-6 C 5.0,1986-09-05 14:25:22,premature ventricular contraction(s). sinus rh...,...,True,NaN,NaN,NaN,NaN,ES,NaN,3,records100/00000/00255_lr,records500/00000/00255_hr
256,565.0,300.0,1,168.0,61.0,8.0,1.0,AT-6 C 5.5,1986-09-05 14:57:27,"premature ventricular contraction(s), bigeminy...",...,True,NaN,NaN,NaN,NaN,NaN,NaN,7,records100/00000/00256_lr,records500/00000/00256_hr
279,437.0,300.0,0,175.0,NaN,6.0,1.0,AT-6 C 5.5,1986-09-09 15:59:30,sinus tachycardia. left axis deviation. left b...,...,True,NaN,NaN,NaN,NaN,NaN,NaN,5,records100/00000/00279_lr,records500/00000/00279_hr
282,437.0,300.0,0,175.0,NaN,3.0,1.0,AT-6 C 5.5,1986-09-10 11:31:08,premature ventricular contraction(s). atrial f...,...,True,NaN,NaN,NaN,NaN,1ES,NaN,5,records100/00000/00282_lr,records500/00000/00282_hr
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
20865,14863.0,300.0,1,NaN,NaN,1.0,2.0,AT-60 3,1999-11-25 07:33:25,sinusrhythmus ueberdrehter linkstyp unspezifis...,...,True,NaN,NaN,NaN,NaN,NaN,NaN,10,records100/20000/20865_lr,records500/20000/20865_hr
20878,11174.0,300.0,1,NaN,NaN,1.0,2.0,AT-60 3,1999-11-28 13:49:12,supraventrikulÄre extrasystole(n) sinusrhythmu...,...,True,NaN,", I-V2,",NaN,NaN,SVES,NaN,9,records100/20000/20878_lr,records500/20000/20878_hr
21121,16658.0,300.0,1,NaN,NaN,1.0,2.0,CS-12,2000-06-20 11:05:26,sinustachykardie lagetyp normal qrs(t) abnorma...,...,True,NaN,", alles,",NaN,NaN,NaN,NaN,6,records100/21000/21121_lr,records500/21000/21121_hr


In [371]:
df.loc[df["age"] == 300, "age"] = 30

In [372]:
df["age"].max()

np.float64(89.0)

In [373]:
df["age"].min()

np.float64(2.0)

In [374]:
df.loc[df["age"] == 2, ["age","height"]]

,age,height
ecg_id,,
10289,2.0,85.0


In [375]:
df["age"] = df["age"] / 89

### sex preprocessing 

In [376]:
df["sex"].value_counts(dropna=False)

sex
0    11354
1    10445
Name: count, dtype: int64

### height preprocessing 

In [377]:
df["height"].value_counts(dropna=False)

height
NaN      14825
160.0      583
165.0      541
170.0      510
168.0      400
         ...  
200.0        1
194.0        1
95.0         1
209.0        1
93.0         1
Name: count, Length: 78, dtype: int64

In [378]:
df["height_missing"] = df["height"].isna().astype(int)

median_height = df["height"].median()

df["height"] = df["height"].fillna(median_height)

In [379]:
df["height"].max()

np.float64(209.0)

In [380]:
df["height"].min()

np.float64(6.0)

In [381]:
(df["height"] == 6).sum()

np.int64(3)

In [382]:
df.loc[df["height"] == 6, "height"] = 60

In [383]:
df["height"].min()

np.float64(60.0)

In [384]:
df["height"] = df["height"] / 209

### weight preprocessing 

In [385]:
df["weight"].value_counts(dropna=False)

weight
NaN      12378
70.0       579
60.0       413
80.0       375
65.0       333
         ...  
33.0         1
200.0        1
183.0        1
128.0        1
131.0        1
Name: count, Length: 128, dtype: int64

In [386]:
df["weight_missing"] = df["weight"].isna().astype(int)

median_weight = df["weight"].median()

df["weight"] = df["weight"].fillna(median_weight)

In [387]:
df["weight"].max()

np.float64(250.0)

In [388]:
df["weight"] = df["weight"] / 250

In [389]:
df.isnull().sum()

patient_id                          0
age                                 0
sex                                 0
height                              0
weight                              0
nurse                            1473
site                               17
device                              0
recording_date                      0
report                              0
scp_codes                           0
heart_axis                       8468
infarction_stadium1             16187
infarction_stadium2             21696
validated_by                     9378
second_opinion                      0
initial_autogenerated_report        0
validated_by_human                  0
baseline_drift                  20201
static_noise                    18539
burst_noise                     21186
electrodes_problems             21769
extra_beats                     19850
pacemaker                       21508
strat_fold                          0
filename_lr                         0
filename_hr 

In [390]:
df["validated_by_human"].value_counts(dropna=False)

validated_by_human
True     16056
False     5743
Name: count, dtype: int64

In [391]:
df["heart_axis"].value_counts(dropna=False)

heart_axis
NaN     8468
MID     7687
LAD     3764
ALAD    1382
RAD      221
ARAD     122
AXL      101
AXR       51
SAG        3
Name: count, dtype: int64

In [392]:
df["infarction_stadium1"].value_counts(dropna=False)

infarction_stadium1
NaN               16187
unknown            3430
Stadium III         980
Stadium II-III      943
Stadium I           166
Stadium II           88
Stadium I-II          5
Name: count, dtype: int64

In [393]:
df["infarction_stadium2"].value_counts(dropna=False)

infarction_stadium2
NaN            21696
Stadium III       65
Stadium I         19
Stadium II        19
Name: count, dtype: int64

In [394]:
df["pacemaker"].value_counts(dropna=False)

pacemaker
NaN              21508
ja, pacemaker      285
ja, nan              3
PACE????, nan        2
?, nan               1
Name: count, dtype: int64

In [395]:
df["extra_beats"].value_counts(dropna=False)

extra_beats
NaN         19850
1ES           405
SVES          381
2ES           179
VES           159
            ...  
1,V1-V6         1
2,V1;V2         1
I-AVR           1
1;V5            1
4ES,SVES        1
Name: count, Length: 129, dtype: int64

In [396]:
df.drop(columns=["nurse", "site","device", "recording_date","validated_by","initial_autogenerated_report","heart_axis"
                 ,"infarction_stadium2","infarction_stadium1","pacemaker","extra_beats","electrodes_problems","burst_noise"
                 ,"static_noise","baseline_drift","second_opinion","filename_lr"], inplace=True)

In [397]:
df.isnull().sum()

patient_id            0
age                   0
sex                   0
height                0
weight                0
report                0
scp_codes             0
validated_by_human    0
strat_fold            0
filename_hr           0
height_missing        0
weight_missing        0
dtype: int64

In [398]:
df["scp_codes"].iloc[4440]

"{'LVH': 100.0, 'ISC_': 100.0, 'IVCD': 100.0, 'LOWT': 0.0, 'STD_': 0.0, 'SR': 0.0}"

In [399]:
df["report"].iloc[7]

'sinusrhythmus linkstyp qrs(t) abnormal    inferiorer infarkt     alter unbest.'

In [400]:
len(df)

21799

In [401]:
from pathlib import Path

BASE = Path(r"D:\AOU-Graduation-Project\BioIntellect\AI\ECG\Data\ptb-xl-a-large-publicly-available-electrocardiography-dataset-1.0.3")

df["filename_hr"] = df["filename_hr"].apply(lambda p: str(BASE / Path(p)))

In [402]:
df["filename_hr"].iloc[0]

'D:\\AOU-Graduation-Project\\BioIntellect\\AI\\ECG\\Data\\ptb-xl-a-large-publicly-available-electrocardiography-dataset-1.0.3\\records500\\00000\\00001_hr'

In [403]:
df.head()

,patient_id,age,sex,height,weight,report,scp_codes,validated_by_human,strat_fold,filename_hr,height_missing,weight_missing
ecg_id,,,,,,,,,,,,
1,15709.0,0.629213,1,0.794258,0.252,sinusrhythmus periphere niederspannung,"{'NORM': 100.0, 'LVOLT': 0.0, 'SR': 0.0}",True,3,D:\AOU-Graduation-Project\BioIntellect\AI\ECG\...,1,0
2,13243.0,0.213483,0,0.794258,0.280,sinusbradykardie sonst normales ekg,"{'NORM': 80.0, 'SBRAD': 0.0}",True,2,D:\AOU-Graduation-Project\BioIntellect\AI\ECG\...,1,0
3,20372.0,0.415730,1,0.794258,0.276,sinusrhythmus normales ekg,"{'NORM': 100.0, 'SR': 0.0}",True,5,D:\AOU-Graduation-Project\BioIntellect\AI\ECG\...,1,0
4,17014.0,0.269663,0,0.794258,0.328,sinusrhythmus normales ekg,"{'NORM': 100.0, 'SR': 0.0}",True,3,D:\AOU-Graduation-Project\BioIntellect\AI\ECG\...,1,0
5,17448.0,0.213483,1,0.794258,0.280,sinusrhythmus normales ekg,"{'NORM': 100.0, 'SR': 0.0}",True,4,D:\AOU-Graduation-Project\BioIntellect\AI\ECG\...,1,0


In [360]:
df.drop(columns=["patient_id", "strat_fold"], inplace=True)

In [404]:
train_folds = [1,2,3,4,5,6,7,8]
val_folds   = [9]
test_folds  = [10]

train_df = df[df["strat_fold"].isin(train_folds)].copy()
val_df   = df[df["strat_fold"].isin(val_folds)].copy()
test_df  = df[df["strat_fold"].isin(test_folds)].copy()

print("train:", train_df.shape, "val:", val_df.shape, "test:", test_df.shape)

train: (17418, 12) val: (2183, 12) test: (2198, 12)


In [405]:
train_df.to_csv("ptbxl_train.csv")
val_df.to_csv("ptbxl_val.csv")
test_df.to_csv("ptbxl_test.csv")

In [407]:
train_patients = set(train_df["patient_id"].dropna().astype(int))
val_patients   = set(val_df["patient_id"].dropna().astype(int))
test_patients  = set(test_df["patient_id"].dropna().astype(int))

print("train ∩ val:", len(train_patients & val_patients))
print("train ∩ test:", len(train_patients & test_patients))
print("val ∩ test:", len(val_patients & test_patients))

train ∩ val: 0
train ∩ test: 0
val ∩ test: 0


### إيه هو الـ Pacemaker؟
جهاز صغير يتزرع جوه الصدر
بيبعت إشارات كهربائية للقلب
علشان يضبط إيقاعه لو القلب بطيء أو متقطع

السبب: القلب مش بيبدأ النبضة لوحده صح

### extra_beats  
دا لو عنده نبضات زياده و لو مش عنده اللي اهو اغلب الداتا ست كدا يبقى هو طبيعي مش عنده حاجه 19850
